ha_gateway.rb

```diff
@@ -27,26 +23,205 @@
 post '/leds' do
   puts params.inspect
 
-  if params['status'] == 'on'
-    api.on
-  elsif params['status'] == 'off'
-    api.off
+  api = LEDENET::Api.new(params['ip'])
+    # Set Power Status
+    if params['status'] == 'on' || params['switch'] == 'on'
+      api.on
+    elsif params['status'] == 'off' || params['switch'] == 'off'
+      api.off
+    end
+
+    r,g,b,w,powerState = api.current_status
+    h,s,l = rgbToHsv(r,g,b)
+
+    # Is the device a Magic UFO?
+    if params['deviceType'].casecmp("UFO").zero?
+      # Adjust HSL only
+      if params.include?('hue') && params.include?('saturation') && params.include?('level')
+        r,g,b = hsvToRgb((params['hue'].to_f*3.6),params['saturation'].to_f,params['level'].to_f)
+        api.update_ufo(r,g,b,w, false)
+      end
+      # Adjust level only
+      if params.include?('level') && !params.include?('hue') && !params.include?('saturation') 
+        h,s,l = rgbToHsv(r,g,b)
+        r,g,b = hsvToRgb(h,s,(params['level'].to_i))
+        api.update_ufo(r,g,b,w, false) 
+      end
+      # Adjust saturation only
+      if params.include?('saturation') && !params.include?('hue') && !params.include?('level') 
+        h,s,v = rgbToHsv(r,g,b)
+        r,g,b = hsvToRgb(h,params['saturation'].to_f,v)
+        api.update_ufo(r,g,b,w, false)
+      end
+      # Adjust hue only
+      if params.include?('hue') && !params.include?('saturation') && !params.include?('level') 
+        h,s,v = rgbToHsv(r,g,b)
+        r,g,b = hsvToRgb(params['hue'].to_f,s,v)
+        api.update_ufo(r,g,b,w, false)
+      end
+      # Adjust Warm White only
+      if params.include?('UFOWWLevel') && !params.include?('saturation') && !params.include?('level') && !params.include?('hue')
+        api.update_ufo(r,g,b,params['UFOWWLevel'].to_i*2.55, false)
+      
+      end
+    else # Adjust a bulb
+      # Adjust HSL
+      if params.include?('hue') && params.include?('saturation') && params.include?('level')
+        # If H & S from ST are the WW button's level, enable WW
+        if params['hue'] == "16.66666666666666" && params['saturation'] == "27.45097875595093"
```

Comment:
SmartThings has a button that sends a value for Warm White in its Color MultiTile. Unfortunately, it doesn't send a direct command to enable WW. The hard-coded values are in place to see if someone pressed the button for WW, and then enables it if if those conditions are true.